### DISCUSSION TITLE
A Problem with GpuIndexIVFPQ

### DISCUSSION THREAD

**Discussion by kamyarrajabalifardi** [facebookresearch/faiss#3791] [Q&A]
Hello everyone!

I'm currently working with **faiss::IndexIVFPQ** and **faiss::GpuIndexIVFPQ** in Python. I would like to change the centroids' embeddings after the training process (centroids of product quantization can be accessed using **index.pq.centroids**).

When I am working with the CPU version (IndexIVFPQ), everything works fine and the search results change when the centroids are updated; however, when I am working with the GPU version, the search results do not change despite the centroids being updated.

My codes are attached below. I would appreciate it if you could help me resolve this issue 🙂
Thanks.

| IVFPQ on CPU | IVFPQ on GPU |
|---------|---------|
| ![Capture](https://github.com/user-attachments/assets/5f59c6fe-c0c6-4751-9ce1-9663df30cfca) | ![Capture](https://github.com/user-attachments/assets/dc2673ef-fe1d-419b-a721-c44c14f6536e) |

**Comment by bmaxdk** [comment]
When you modify the centroids on the CPU, these changes are not automatically reflected on the GPU. The search operation still uses the centroids stored on the GPU, which have not been updated.

after this line, `faiss.copy_array_to_vector(new_centroids, index.pq.centroids)`, ' can you add `index.copyToGpu(index_pq)`? It will synchronize the centroid to the GPU
